scripts/verify_commits

```diff
@@ -0,0 +1,133 @@
+#!/bin/bash
+
+function err() {
+    echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $*" >&2
+}
+
+function verify_staging_diff() {
+    local remote="${1}"
+    local upstream_commit="${2}"
+    local downstream_commit="${3}"
+    local staging_dir="staging/${remote}"
+
+    local outside_staging
+    outside_staging="$(git show --name-only "${downstream_commit}" -- ":!${staging_dir}")"
+    if [[ -n "${outside_staging}" ]]; then
+        err "downstream staging commit ${downstream_commit} changes files outside of ${staging_dir}:"
+        err "${outside_staging}"
+        err "factor out changes to ${staging_dir} into a standalone commit"
+        return 1
+    fi
+
+    local divergence
+    divergence="$(git diff --stat "${upstream_commit}" "${downstream_commit}:${staging_dir}" -- ':!vendor')"
+    if [[ -n "${divergence}" ]]; then
+        err "downstream staging commit ${downstream_commit} diverges from upstream commit ${upstream_commit}:"
+        err "${divergence}"
+        err "changes must match referenced upstream commit ${upstream_commit}"
+        return 1
+    fi
+}
+
+function verify_root_diff() {
+    local downstream_commit="${1}"
+
+    local inside_staging
+    inside_staging="$(git show --name-only "${downstream_commit}" -- staging)"
+    if [[ -n "${inside_staging}" ]]; then
+        err "downstream non-staging commit ${downstream_commit} changes staging"
+        err "${inside_staging}"
+        err "only staging commits (i.e. from an upstream cherry-pick) may change staging"
+
+        return 1
+    fi
```

Comment:
we should be able to make downstream-only changes to staging, so should I change this to check for something else? Maybe "doesn't make changes to more than one staging directory at once"?